src/deploy/functions/containerCleaner.ts

```diff
@@ -128,6 +129,112 @@ export class ContainerRegistryCleaner {
   }
 }
 
+function getHelper(cache: Record<string, DockerHelper>, subdomain: string): DockerHelper {
+  if (!cache[subdomain]) {
+    cache[subdomain] = new DockerHelper(`https://${subdomain}.${containerRegistryDomain}`);
+  }
+  return cache[subdomain];
+}
+
+/**
+ * List all paths from the GCF directory in GCR (e.g. us.gcr.io/project-id/gcf/location).
+ * @param projectId: the current project that contains GCF artifacts
+ * @param location: the specific region to search for artifacts. If omitted, will search all locations.
+ * @param dockerHelpers: a map of {@link SUBDOMAINS} to {@link DockerHelper}. If omitted, will use the default value and create each {@link DockerHelper} internally.
+ *
+ * @throws {@link FirebaseError}
+ * Thrown if the provided location is not a valid Google Cloud region or we fail to search subdomains.
+ */
+export async function listGcfPaths(
+  projectId: string,
+  locations?: string[],
+  dockerHelpers: Record<string, DockerHelper> = {}
+): Promise<string[]> {
+  if (!locations) {
+    locations = Object.keys(SUBDOMAIN_MAPPING);
+  }
+  const invalidRegion = locations.find((loc) => !SUBDOMAIN_MAPPING[loc]);
+  if (invalidRegion) {
+    throw new FirebaseError(`Invalid region ${invalidRegion} supplied`);
+  }
+  const locationsSet = new Set(locations); // for quick lookup
+  const subdomains = new Set(Object.values(SUBDOMAIN_MAPPING));
+  const failedSubdomains = new Set<string>();
+
+  const listAll = [...subdomains].map((subdomain) => {
```

Comment:
Looping through the subdomains here and filtering later since that limits us to calling the docker API only 3 times.